### DISCUSSION TITLE
setEiData

### DISCUSSION THREAD

**Discussion by GianmariaTS** [fattureincloud/api#244] [Q&A] [answered]
nella migrazione delle API da v1 a v2 sto cercando di caricare la fattura elettronica PA e mi trovo bloccato con la funzione setEiData in ambito php utilizzando il seguente codice:

`
// Here we set e_invoice and ei_data
$invoice->setEInvoice(true);
$invoice->setEiData(
    new IssuedDocumentEiData(
        array(
            "payment_method" => "MP05"
        )
    )
);
` 
l'esecuzione si blocca quando viene chiamata la funzione setEiData

Da cosa potrebbe dipendere?

Grazie,
Gian

**Comment by valmoz** [comment]
Buongiorno @GianmariaTS 
cosa intende per "l'esecuzione si blocca"? ha qualche errore?

Grazie
Mauro

**Reply by GianmariaTS** [reply]
non mi da errore... esce senza eseguire le altre righe di codice

**Comment by GianmariaTS** [comment]
non mi da errore... esce senza eseguire le altre righe di codice. 
Fornisco il codice di esempio completo:




require_once(__DIR__ . '/vendor/autoload.php');


use FattureInCloud\Model\Currency;
use FattureInCloud\Model\DocumentTemplate;
use FattureInCloud\Model\Entity;
use FattureInCloud\Model\IssuedDocument;
use FattureInCloud\Model\IssuedDocumentItemsListItem;
use FattureInCloud\Model\IssuedDocumentPaymentsListItem;
use FattureInCloud\Model\IssuedDocumentStatus;
use FattureInCloud\Model\IssuedDocumentType;
use FattureInCloud\Model\CreateIssuedDocumentRequest;
use FattureInCloud\Model\Language;
use FattureInCloud\Model\PaymentAccount;
use FattureInCloud\Model\PaymentMethod;
use FattureInCloud\Model\VatType;




// Retrieve the access token from the session variable
$accessToken = 'xxx'; 

// Get the API config and construct the service object.
$config = FattureInCloud\Configuration::getDefaultConfiguration()->setAccessToken($accessToken);

$userApi = new FattureInCloud\Api\UserApi(
    new GuzzleHttp\Client(),
    $config
);
$suppliersApi = new FattureInCloud\Api\SuppliersApi(
    new GuzzleHttp\Client(),
    $config
);

$companies = $userApi->listUserCompanies();
$company_id = $companies->getData()->getCompanies()[0]->getId();
echo $company_id ;
// var_dump($companies);

$apiInstance = new FattureInCloud\Api\IssuedDocumentsApi(
    new GuzzleHttp\Client(),
    $config
);
//set your company id
    
$entity = new Entity;
$entity
    ->setId(3)
    ->setName("xxx")
    ->setVatNumber("xxx")
    ->setTaxCode("xxx")
    ->setAddressStreet("xxx")
    ->setAddressPostalCode("xxx")
    ->setAddressCity("xxx")
    ->setAddressProvince("xx")
    ->setCountry("Italia");

$invoice = new IssuedDocument;
$invoice->setType(IssuedDocumentType::INVOICE);
$invoice->setEntity($entity);
$invoice->setDate(new DateTime("2023-01-01"));
$invoice->setNumber(999);
$invoice->setNumeration("/fatt");
$invoice->setSubject("internal subject");
$invoice->setVisibleSubject("visible subject");
$invoice->setCurrency(
    new Currency(
        array(
           "id" => "EUR" 
        )
    )
);
$invoice->setLanguage(
    new Language(
        array(
            "code" => "it",
            "name" => "italiano"
        )
    )
);
$invoice->setItemsList(
    array(
        new IssuedDocumentItemsListItem(
            array(
                "product_id" => 4,
                "code" => "TV3",
                "name" => "prova",
                "net_price" => 100,
                "category" => "cucina",
                "discount" => 0,
                "qty" => 1,
                "vat" => new VatType(
                    array(
                        "id" => 0
                    )
                )
            )
        )
    )
);

// Here we set e_invoice and ei_data
$invoice->setEInvoice(true);
$invoice->setEiData(
    new IssuedDocumentEiData(
        array(
            "payment_method" => "MP05"
        )
    )
);


$invoice->setPaymentMethod(
    new PaymentMethod(
        array(
            "id" => 1289340
        )
    )
);

$invoice->setPaymentsList(
    array(
        new IssuedDocumentPaymentsListItem(
            array(
                "amount" => 122,
                "due_date" => new DateTime("2023-01-23"),
                "paid_date" => new DateTime("2023-01-22"),
                "status" => IssuedDocumentStatus::PAID,
                "payment_account" => new PaymentAccount(
                    array(
                        "id" => 268627
                    )
                )
            )
        )
    )
);

// Here we put our invoice in the request object
$create_issued_document_request = new CreateIssuedDocumentRequest();
$create_issued_document_request->setData($invoice);

// Now we are all set for the final call
// Create the invoice: https://github.com/fattureincloud/fattureincloud-php-sdk/blob/master/docs/Api/IssuedDocumentsApi.md#createissueddocument
try {
    $result = $apiInstance->createIssuedDocument($company_id, $create_issued_document_request);
    print_r($result);
} catch (Exception $e) {
    echo 'Exception when calling IssuedDocumentsApi->createIssuedDocument: ', $e->getMessage(), PHP_EOL;
}

**Comment by emajo** [comment, marked as answer]
Ciao @GianmariaTS, sembrerebbe mancare l'import della classe IssuedDocumentEiData, dovrebbe bastare aggiungerlo facendo 
`use FattureInCloud\Model\IssuedDocumentEiData;`

Ci faccia sapere se così risolve,
grazie,
Emanuele.

**Reply by GianmariaTS** [reply]
Ciao @emajo hai perfettamente ragione, il problema era quello. Grazie

**Reply by GianmariaTS** [reply]
@emajo posso chiederti se sai come posso valorizzare il codice destinatario? Non riesco a trovarlo... Grazie

**Reply by emajo** [reply]
Il codice destinatario è `entity.ei_code`, abbiamo una guida sulla creazione della fattura che spiega gran parte dei campi utili [qui](https://developers.fattureincloud.it/docs/guides/invoice-creation).

**Reply by GianmariaTS** [reply]
Grazie @emajo
